rosetta/services/operation_components.go

```diff
@@ -0,0 +1,235 @@
+		components.To = op1.Account
+	} else {
+		components.From = op1.Account
+		components.To = op0.Account
+	}
+	if components.From == nil || components.To == nil {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": "both operations must have account identifiers for same shard transfer",
+		})
+	}
+	return components, nil
+}
+
+// getCrossShardOperationComponents ..
+func getCrossShardOperationComponents(
+	operation *types.Operation,
+) (*OperationComponents, *types.Error) {
+	if operation == nil {
+		return nil, common.NewError(common.CatchAllError, map[string]interface{}{
+			"message": "nil operation",
+		})
+	}
+	metadata := common.CrossShardTransactionOperationMetadata{}
+	if err := metadata.UnmarshalFromInterface(operation.Metadata); err != nil {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": errors.WithMessage(err, "invalid metadata").Error(),
+		})
+	}
+	amount, err := types.AmountValue(operation.Amount)
+	if err != nil {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": err.Error(),
+		})
+	}
+	if amount.Sign() == 1 {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": "sender amount must not be positive for cross shard transfer",
+		})
+	}
+	if types.Hash(operation.Amount.Currency) != common.CurrencyHash {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": "invalid currency for provided amounts",
+		})
+	}
+
+	components := &OperationComponents{
+		Type:   operation.Type,
+		To:     metadata.To,
+		From:   metadata.From,
+		Amount: new(big.Int).Abs(amount),
+	}
+	if components.From == nil || components.To == nil || operation.Account == nil {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": "operation must have account sender/from & receiver/to identifiers for cross shard transfer",
+		})
+	}
+	if types.Hash(operation.Account) != types.Hash(components.From) {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": "operation account identifier does not match sender/from identifiers for cross shard transfer",
+		})
+	}
+	return components, nil
+}
+
+// getContractCreationOperationComponents ..
+func getContractCreationOperationComponents(
+	operation *types.Operation,
+) (*OperationComponents, *types.Error) {
+	if operation == nil {
+		return nil, common.NewError(common.CatchAllError, map[string]interface{}{
+			"message": "nil operation",
+		})
+	}
+	amount, err := types.AmountValue(operation.Amount)
+	if err != nil {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": err.Error(),
+		})
+	}
+	if amount.Sign() == 1 {
```

Comment:
for contract creation, you could still send token along to the contract address right? By making the contract constructor payable.

Reply:
correct, but then the amount should be **negative** for the rosetta spec as it would deduct funds from the contract deployer.